# 🚀 Project 1: Customer Churn & Retention Analysis
### *A Step-by-Step, Beginner-Friendly Journey into Real-World Data Science & Machine Learning*

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/AsemaniJoshua/Retention-Analysis-Data-Science-Project/blob/main/telco_customer_churn_analysis.ipynb)
![Python](https://img.shields.io/badge/Python-3.9+-blue.svg)
![Dataset](https://img.shields.io/badge/Dataset-IBM%20Telco%20Churn-orange.svg)
![Status](https://img.shields.io/badge/Section%201-Completed-success.svg)

---

### 📖 Welcome to Your Data Science Journey!
Imagine you run an ice cream club with your friends. Every month, members pay $5 to taste brand new flavors.
* Some friends **stay** month after month because they love the ice cream and service.
* But other friends **quit and leave** because the ice cream melted, or it was too expensive, or the line was too long!

In the business world, when a customer decides to leave and cancel their subscription, we call that **"Churn"** (rhymes with *burn*).
Think of your business as a bucket of water:
* Winning new customers pours fresh water into the bucket.
* Customers leaving is a **hole in the bottom** where water drips out.
* If the leak is too big, the bucket will soon be empty, no matter how much water you pour in!

**Our Mission as Data Detectives:**
We are going to use math, python code, and machine learning to find out:
1. **Who is about to leave?** (Prediction)
2. **Why are they leaving?** (Root cause & feature importance)
3. **What can the company do to make them stay?** (Retention strategy)

---

## 🗺️ Project Roadmap
- [x] **Section 1: Setup, Ingestion & Initial Detective Work** *(We are here!)*
- [ ] **Section 2: Data Cleaning & Column Dictionary (Child-Friendly Guide)**
- [ ] **Section 3: Exploratory Data Analysis (EDA) & Storytelling Visualizations**
- [ ] **Section 4: Feature Engineering & Preprocessing Pipeline**
- [ ] **Section 5: Handling Class Imbalance (SMOTE vs. Class Weights)**
- [ ] **Section 6: Model Training & Head-to-Head Comparison (Logistic Regression, Random Forest, XGBoost)**
- [ ] **Section 7: Model Evaluation & The ROC-AUC Curve Explained Simply**
- [ ] **Section 8: Feature Importance, Retention Strategies & Executive Recommendations**

---


## 🔬 Section 1: Setup, Ingestion & Initial Detective Work

In this first section, we will:
1. **Gather our tools**: Import the special Python libraries we need and explain what each one does like you're explaining it to a 5-year-old.
2. **Load our dataset**: Pull the live IBM Telco data directly from the internet.
3. **Take our first look**: Inspect the size, rows, and columns of the data.
4. **Do some detective work**: Spot an invisible bug in the dataset that trips up even professional data scientists!


### 🛠️ Step 1.1: Meet the Toolkit (Libraries)

Think of Python as your Swiss Army Knife, and **libraries** are the special gadgets attached to it:
* **`pandas`** 🐼: Think of this as a supercharged digital Excel spreadsheet on steroids. It lets us hold tables of data, filter rows, rename columns, and calculate sums in a split second.
* **`numpy`** 🔢: The lightning-fast math wizard. It does heavy-duty calculations with numbers, arrays, and matrices behind the scenes.
* **`matplotlib.pyplot`** 🎨: Our blank canvas and digital colored pencils. We use it to draw charts, graphs, and plots.
* **`seaborn`** 🌊: A fancy artist friend of Matplotlib who automatically makes charts look gorgeous, modern, and easy to read.
* **`warnings`** 🔕: A small tool to tell Python: *"Please don't show annoying red warning messages if everything is working fine!"*

Let's run the code below to bring all these tools into our notebook!


In [1]:
# ==============================================================================
# STEP 1.1: IMPORT OUR ESSENTIAL DATA SCIENCE LIBRARIES
# ==============================================================================

# 1. Pandas: Our table organizer (like Excel inside Python code)
import pandas as pd

# 2. NumPy: Our lightning-fast math calculator for numbers and lists
import numpy as np

# 3. Matplotlib: Our digital canvas for drawing charts and plots
import matplotlib.pyplot as plt

# 4. Seaborn: Our stylist that makes Matplotlib charts look modern and pretty
import seaborn as sns

# 5. Warnings: To suppress harmless version update alerts so our notebook stays clean
import warnings
warnings.filterwarnings('ignore')

# 6. Make our charts look clean with a crisp whitegrid theme
sns.set_theme(style="whitegrid", palette="muted")
plt.rcParams['figure.figsize'] = (10, 5) # Default width and height for charts
plt.rcParams['figure.dpi'] = 100         # High-resolution, sharp visual output

print("[SUCCESS] All core libraries imported and ready for action!")


[SUCCESS] All core libraries imported and ready for action!
[SUCCESS] All core libraries imported and ready for action!


### 📥 Step 1.2: Loading the Dataset Directly from the Cloud

Instead of downloading a file to your computer and manually uploading it, we can tell Pandas to download it directly from a web link (URL).

* **Where is this data from?** It is IBM's famous **Telco Customer Churn** dataset, hosted publicly on GitHub.
* **What is a CSV?** CSV stands for *Comma-Separated Values*. It is just plain text where every row is a person, and their details are separated by commas (`,`).


In [2]:
# ==============================================================================
# STEP 1.2: DOWNLOAD & LOAD THE DATASET INTO PANDAS
# ==============================================================================

# The public web address where the raw CSV file lives on GitHub
dataset_url = "https://raw.githubusercontent.com/IBM/telco-customer-churn-on-icp4d/master/data/Telco-Customer-Churn.csv"

# Tell pandas to read the CSV directly from the internet and store it in a variable called 'df'
# 'df' stands for DataFrame (our digital table)
df = pd.read_csv(dataset_url)

# Print a friendly message to confirm it worked!
print("[SUCCESS] Dataset successfully downloaded and loaded into memory!")
print(f"[INFO] Our dataset has {df.shape[0]:,} customer rows and {df.shape[1]} clue columns.")


[SUCCESS] Dataset successfully downloaded and loaded into memory!
[INFO] Our dataset has 7,043 customer rows and 21 clue columns.
[INFO] Our dataset has 7,043 customer rows and 21 clue columns.


### 🔍 Step 1.3: Inspecting the First 5 Rows

Now let's ask Pandas to show us the top 5 rows using `.head()`.
* Each **horizontal row** is one single customer (a real person who subscribed to phone/internet).
* Each **vertical column** is a piece of information (a "feature" or "clue") about that person:
  - Are they male or female?
  - How many months have they stayed with us (`tenure`)?
  - What services do they pay for (Internet, TV, Tech Support)?
  - How much do they pay each month (`MonthlyCharges`)?
  - Did they leave the company (`Churn`: Yes or No)?


In [3]:
# ==============================================================================
# STEP 1.3: PREVIEW THE FIRST 5 CUSTOMERS IN OUR TABLE
# ==============================================================================

# .head(5) means: "Hey pandas, show me the top 5 customers from the table!"
df.head(5)


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


### 🩻 Step 1.4: The Data X-Ray (`.info()`)

Just like a doctor takes an X-ray to look inside a patient, a data scientist runs `df.info()` to see what types of data are inside the table:
* **`int64`**: Whole numbers (like `SeniorCitizen`: 0 for No, 1 for Yes).
* **`float64`**: Numbers with decimals (like `MonthlyCharges`: $29.85).
* **`object`**: Text / Words (like `gender`: "Female", `Partner`: "Yes").

🚨 **Wait, look closely at the output below!** Look at `TotalCharges`. Total charges is money, so it should be a decimal number (`float64`), right? Why does it say **`object` (text)**?
Let's find out!


In [4]:
# ==============================================================================
# STEP 1.4: CHECK DATA TYPES AND MISSING VALUES
# ==============================================================================

# .info() prints summary information about column names, non-empty counts, and data types
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

### 🕵️ Step 1.5: The Detective Work — Finding the "Invisible Sneak"

Why did Pandas label `TotalCharges` as text (`object`) instead of a number?
Because inside the `TotalCharges` column, there are sneaky rows that do not contain numbers... they contain an **empty blank space** (`" "`)!

Imagine someone asked you to write your phone bill on a test paper, and you just drew an empty blank box. The computer doesn't know that empty box means `$0.00`; the computer thinks: *"Hey, that's a piece of text (a space character), so the whole column must be text!"*

Let's find out:
1. How many of these blank space rows exist?
2. Who are these customers, and why are their charges blank?


In [5]:
# ==============================================================================
# STEP 1.5: UNCOVERING THE SNEAKY BLANK SPACES IN 'TotalCharges'
# ==============================================================================

# Count how many rows have an exact blank space ' ' as their TotalCharges
blank_space_count = (df['TotalCharges'] == ' ').sum()

print(f"[FOUND] Detective finding: We found exactly {blank_space_count} rows with blank spaces in 'TotalCharges'!")

# Let's inspect these 11 mystery customers!
# We filter the dataframe to only look at rows where TotalCharges is equal to a single space ' '
mystery_customers = df[df['TotalCharges'] == ' ']

# Let's display their Customer ID, their tenure (months with company), Monthly Charges, and Total Charges
mystery_customers[['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges', 'Churn']]


[FOUND] Detective finding: We found exactly 11 rows with blank spaces in 'TotalCharges'!


,customerID,tenure,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,0,52.55,,No
753,3115-CZMZD,0,20.25,,No
936,5709-LVOEQ,0,80.85,,No
1082,4367-NUYAO,0,25.75,,No
1340,1371-DWPAZ,0,56.05,,No
3331,7644-OMVMY,0,19.85,,No
3826,3213-VVOLG,0,25.35,,No
4380,2520-SGTTA,0,20.00,,No
5218,2923-ARZLG,0,19.70,,No
6670,4075-WKNIU,0,73.35,,No


#### 💡 The Mystery Solved!
Look at the `tenure` column for all 11 mystery customers:
**Every single one of them has `tenure = 0`!**

What does `tenure = 0` mean?
* They **just joined the company this month**! They are brand new baby customers!
* Because they just joined today, they haven't received their first monthly bill yet.
* That is why their `TotalCharges` was left blank!

In **Section 2**, we will cleanly fix this by converting this column to real numbers and setting their `TotalCharges` to `0.0` or imputing with the median.


### 📊 Step 1.6: Quick Summary Statistics (`.describe()`)

Now let's ask Pandas to summarize the numbers in our dataset:
* **`count`**: How many people have numbers in this column.
* **`mean`**: The arithmetic average (add everyone up and divide by total people).
* **`std`** (Standard Deviation): How spread out the numbers are. (Are all people paying roughly the same, or do some pay $20 and others pay $120?).
* **`min`**: The lowest number.
* **`25%`, `50%`, `75%`**: The percentiles (50% is the **Median**, meaning half the people pay less, half pay more).
* **`max`**: The highest number.


In [6]:
# ==============================================================================
# STEP 1.6: SUMMARY STATISTICS OF NUMERICAL COLUMNS
# ==============================================================================

# .describe() computes statistical summaries for all numeric columns
df.describe()


,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


#### 📝 What do these numbers tell us already?
1. **`tenure`** (Months as a customer):
   - Average tenure is **~32 months** (almost 3 years!).
   - Minimum is **0 months** (the new subscribers we just discovered).
   - Maximum is **72 months** (loyal veteran customers who have been here for 6 full years!).
2. **`MonthlyCharges`**:
   - The cheapest plan is **$18.25/month** (probably basic phone only).
   - The average plan is **$64.76/month**.
   - The most expensive plan is **$118.75/month** (probably high-speed fiber internet + movies + phone + tech support).


---
### 🎯 Section 1 Summary & Checkpoint: What Did We Accomplish?

| 🏅 Checkpoint | What We Learned / Did | Why It Matters for Data Science |
| :--- | :--- | :--- |
| **Libraries** | Imported `pandas`, `numpy`, `matplotlib`, `seaborn` | Gave us the tools to manipulate data and draw charts |
| **Data Ingestion** | Loaded 7,043 rows and 21 columns directly via URL | Fully reproducible in Google Colab without manual file uploads |
| **Inspection** | Used `.head()`, `.info()`, and `.describe()` | Understand our rows (customers) and columns (attributes) |
| **Detective Work** | Found 11 rows with `" "` in `TotalCharges` (all `tenure=0`) | Uncovered a hidden data bug before it crashes our machine learning models! |

---
👉 **Next Up: Section 2 (Data Cleaning & Column Dictionary)**
In the next section, we will:
1. Transform `TotalCharges` into clean numeric floats and fix the 11 blank spaces.
2. Build a **Child-Friendly Column Dictionary** that walks through all 21 columns in plain English!
3. Drop unnecessary identification columns (`customerID`) so our machine learning model doesn't get distracted.
